modules/compatcheck/main.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"context"
+	"dagger/compatcheck/internal/dagger"
+	"fmt"
+	"runtime"
+
+	"github.com/dagger/dagger/dagql/introspection"
+	"github.com/josephburnett/jd/v2"
+	"github.com/moby/buildkit/identity"
+	"github.com/tidwall/gjson"
+	"golang.org/x/exp/rand"
+	"golang.org/x/mod/semver"
+)
+
+type Compatcheck struct{}
+
+// Compare the schema of given module with different versions of engine
+func (m *Compatcheck) Run(ctx context.Context,
+	// ref of the module to compare
+	module,
+	// version of engine to compare with
+	versionA,
+	// version of engine to compare to
+	versionB string,
+	//+optional
+	// only required when one of the version to compare is 'dev' [optional]
+	source *dagger.Directory) error {
+	if (versionA == "dev" || versionB == "dev") && source == nil {
+		return fmt.Errorf("--source flag is required when one of the requested engine version is 'dev'")
+	}
+
+	schemaA, err := m.getSchemaForModuleForEngineVersion(ctx, module, versionA, source)
+	if err != nil {
+		return err
+	}
+
+	schemaB, err := m.getSchemaForModuleForEngineVersion(ctx, module, versionB, source)
+	if err != nil {
+		return err
+	}
+
+	a, _ := jd.ReadJsonString(schemaA)
+	b, _ := jd.ReadJsonString(schemaB)
+
+	diff := a.Diff(b).Render()
+	if diff != "" {
+		return fmt.Errorf("%s", diff)
+	}
+
+	return nil
+}
+
+// setup dagger engine/client with requested version and
+// fetches schema using dagger query
+func (m *Compatcheck) getSchemaForModuleForEngineVersion(ctx context.Context, module, engineVersion string, source *dagger.Directory) (string, error) {
+	var engineSvc *dagger.Service
+	var client *dagger.Container
+	var err error
+
+	if engineVersion == "dev" {
+		client = devEngineAndClient(source)
+	} else {
+		engineSvc = engineServiceWithVersion(engineVersion)
+		client, err = engineClientContainerWithVersion(ctx, engineSvc, engineVersion)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	rawIntrospection, err := client.WithNewFile("/schema-query.graphql", introspection.Query).
+		WithExec([]string{"dagger", "query", "-m", module, "--doc", "/schema-query.graphql"}).
+		Stdout(ctx)
+
+	if err != nil {
+		return "", err
+	}
+
+	return gjson.Get(rawIntrospection, "__schema").String(), nil
```

Comment:
As we just discussed, we can avoid accidentally comparing *all* the core APIs between versions (which we expect to change frequently) by first taking a baseline introspection query without the module loaded, and then removing all the `types` that are present in both this baseline and the module schema.

Then we can just end up diffing the module introspection.

Reply:
Hi Justin, 

thank you for the review and your time to discuss this yesterday.

I implemented a `schemadiff` package for it, which basically do what we discussed, but slightly differently.

Quoting from the doc that I added to the `schemadiff` package:

```
// This function takes four params:
// 1. base schema with version-a of engine
// 2. base schema with version-b of engin
// 3. schema with requested module installed with version-a of engine
// 4. schema with requested module installed with version-b of engine
//
// It first calculate the diff between base schema's of version-a and version-b
// and then calculate the diff between full schema (with module installed) of version-a and version-b.
// Then it verifies if the full schema diff has more changes after ignoring the base schema diff,
// which essentially is the schema diff in the module schema.
//
// In order to ignore that diff from base schema, it normalizes a couple of things in the jd.DiffElement object:
// 1. It replaces all the jd.PathElement of type jd.PathIndex with constant value of 0 (basically ignoring index)
// 2. It removes the Before and After context from the diff. This context is useful when rendering the diff but ignored here.
```

I tested with a few modules, and it seems to be working as expected, though we may get feedback when this is used in real world scenarios and this module find real (or false positive) diff. 

As a next step, I think it will be useful to add an option to consider the schema diff in base (e.g. when checking for base schema during dagger development itself).